### ISSUE TITLE
enveloped-signature must go first for verification to work properly

### ISSUE THREAD

**Issue by joeesteves** [node-saml/xml-crypto#111]
Hi, i was dealing with a verifying problem. 
`[ 'invalid signature: for uri #_0 calculated digest is I6QnT6OKRlqeJXVGIb/UIiV1k/ArtW5/fAJkjGoxP8Y= but the xml to validate supplies digest 3g7fAmI5/vWlkDLK4zlU3/GSmjdFOlOCGs+pNMNqSVM=' ]`
The calculated digest when verifying was different from the generated when signing.  The problem i had was the order of the transform algorithm. 
Have to change mi code from

``` javascript
beforeChanges :

sig.addReference("//*[local-name(.)='CFE']", [ "http://www.w3.org/2001/10/xml-exc-c14n#","http://www.w3.org/2000/09/xmldsig#enveloped-signature"])

afterChanges: 
sig.addReference("//*[local-name(.)='CFE']", ["http://www.w3.org/2000/09/xmldsig#enveloped-signature", "http://www.w3.org/2001/10/xml-exc-c14n#"])
```

The problem with the first code, was that exclusive-canonicalization algorithm returns a string and enveloped-signature can't find signature form string, it must recibe node format to work ok.  This can be repaired adding one line of code on enveloped-signature.js

``` javascript

EnvelopedSignature.prototype.process = function (node) {
  node = new Dom().parseFromString(node)  // new line
  var signature = xpath(node, ".//*[local-name(.)='Signature' and namespace-uri(.)='http://www.w3.org/2000/09/xmldsig#']")[0];
  if (signature) signature.parentNode.removeChild(signature);
  return node;
};
```

Not sure if it would be wright to support both transformation orders ([ "http://www.w3.org/2001/10/xml-exc-c14n#","http://www.w3.org/2000/09/xmldsig#enveloped-signature"] or [ "http://www.w3.org/2000/09/xmldsig#enveloped-signature","http://www.w3.org/2001/10/xml-exc-c14n#"])

**Comment by markstos** [comment]
I think you are on the right track based on this spec: https://www.w3.org/TR/xmldsig-core/ in 6.6.3 XPath Filtering

> Each signature must omit itself from its own digest calculations, but it is also necessary to exclude the second signature element from the digest calculations of the first signature so that adding the second signature does not break the first signature.

**Comment by ponyesteves** [comment]
Yes @markstos I agree with that. The doubt I have is, say you have this code
```xml
<xmltosign>
<data>....</data>
<signature>....</signature>
</xmltosign>
When verifiying, the envelope-signature transformation indicates to remove signature node to before calculate the document digest, and the second transformation will indicate to normalice to exc-c14n document. I think this is the correct order and in fact it's what works properly with this library. But what I can't confirm is if it is a regulated thing by the standard (can't find it in the official documentation) or it's simple the logical order.  To sum up xml-crypto works fine with first order of transformation and fails to verify with the second order. If the order is a standardized thing perhaps an error message would save lot of time for those that get stucked on this. 
I've already revolve the verifying stage but I have a hard time figuring the errors I was making and any error message would have been useful.   
I'd be glad to help if is a valid requirement.

**Comment by cjbarth** [comment]
@ponyesteves , if you'd be willing to contribute a PR with a test suite, even if that PR is just a better error message, we'd love to have it. Please consider helping the community in this way.

**Comment by cjbarth** [comment]
This still reproduces on `master` (4359e84) and on 6.1.2. It is broader than `enveloped-signature`: `getCanonXml` stops applying transforms as soon as one returns a string, so every transform after a canonicalization is silently skipped.

```ts
const sig = new SignedXml({
  privateKey,
  canonicalizationAlgorithm: "http://www.w3.org/2001/10/xml-exc-c14n#",
  signatureAlgorithm: "http://www.w3.org/2001/04/xmldsig-more#rsa-sha256",
});
sig.addReference({
  xpath: "/*",
  transforms: [
    "http://www.w3.org/2001/10/xml-exc-c14n#",
    "http://www.w3.org/2000/09/xmldsig#enveloped-signature",
  ],
  digestAlgorithm: "http://www.w3.org/2001/04/xmlenc#sha256",
});
sig.computeSignature("<root><x>1</x></root>");
// loading that signature and calling checkSignature(sig.getSignedXml()) returns false
```

- `[exc-c14n, enveloped-signature]` on the root: signing succeeds, but xml-crypto cannot verify its own signature. The signer digests before the `Signature` exists; the verifier skips `enveloped-signature` and digests the `Signature` too. It fails closed.
- `[exc-c14n, <custom transform>]`: `process()` is never called on either side, so the signature verifies with xml-crypto while `SignedInfo` declares a transform that was never applied. A conforming verifier computes a different digest.

The order is legal. [XMLDSig 4.4.3.2](https://www.w3.org/TR/xmldsig-core1/#sec-ReferenceProcessingModel) says that when the data is an octet stream and the next transform requires a node-set, the application MUST parse the octets into one. The enveloped signature transform ([6.6.4](https://www.w3.org/TR/xmldsig-core1/#sec-EnvelopedSignature)) is such a transform. .NET does this: [`XmlDsigEnvelopedSignatureTransform`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/XmlDsigEnvelopedSignatureTransform.cs) accepts a `Stream`, loads it into an `XmlDocument`, and finds the signature by position.

So the fix suggested above is the right rule, but it belongs in `getCanonXml`, applied between any two transforms, rather than inside `EnvelopedSignature`. One thing to design for: after re-parsing, `options.signatureNode` no longer points into the tree, so `EnvelopedSignature` has to rely on its existing `SignatureValue` match.

This fits 6.2 as a bug fix. Chains with `enveloped-signature` after a canonicalization fail today whenever the `Signature` is inside the referenced element, which is the case that transform exists for, so those only start verifying. Chains where a custom transform follows a canonicalization get a different digest, which needs an Upgrading note like the one added in #581. Throwing on such chains instead would reject configurations 6.x accepts.
